src/azure-cli-telemetry/azure/cli/telemetry/components/records_collection.py

```diff
@@ -35,8 +37,11 @@ def snapshot_and_read(self):
         if not os.path.isdir(folder):
             return
 
+        # Collect all cache.x files. If it has been a long time since last sent, also collect cache file itself.
+        include_cache = datetime.datetime.now() - self._last_sent > datetime.timedelta(hours=self._get_threshold_config())
```

Comment:
We should determine whether we want 

- since last sent, which compares `TELEMETRY_NOTE_NAME`'s content: https://github.com/Azure/azure-cli/blob/be601b7cfe542358f954e180b61b02c98fcfb27b/src/azure-cli-telemetry/azure/cli/telemetry/components/telemetry_note.py#L40-L41, or
- since last check, which compares `TELEMETRY_NOTE_NAME`'s `st_mtime` https://github.com/Azure/azure-cli/blob/be601b7cfe542358f954e180b61b02c98fcfb27b/src/azure-cli-telemetry/azure/cli/telemetry/util.py#L34-L38

Reply:
To clarify, last sent is the record time of last uploaded, last check is the time we check if there's any cache need to be uploaded. So last check can be updated every 10 mins even there's no cache uploaded.

Think about this scenario if we use last check time to compare:
1. `07-14 15:00`, cx runs a cli command. A new record has been stored in cache file while the file size is 6k.
-  Since no telemetry uploaded, so last sent is the time of previous record, let's say it's `07-14 14:30`.
- Last check has been updated to `07-14 15:00`.
3. Then `07-15 14:40`, cs runs another command. Another record has been stored in cache file and the file size grows to 9k.
-  now(`07-15 14:40`) - last check time(`07-14 15:00`) is less than 24 hours, so cache file won't be uploaded
- last sent time is still `07-14 14:30`, last check time is updated to `07-15 14:40`
5. Another 10 mins later, `07-15 14:50`, cs runs another command. Another record has been stored in cache file and the file size grows to 12k.
- now(`07-15 14:50`) - last check time(`07-15 14:40`) is less than 24 hours, so cache file won't be uploaded
- last sent time is still `07-14 14:30`, last check time is updated to `07-15 14:50`
......
......
If customer keeps running cli commands, last check time will keep updating, cache file will never be uploaded until rotated with cache.1

Reply:
Very good analysis! Thanks for clarifying it. 😊